metaflow/metaflow_git.py

```diff
@@ -0,0 +1,131 @@
+#!/usr/bin/env python
+"""Get git repository information for the package
+
+Functions to retrieve git repository details like URL, branch name, 
+and commit SHA for Metaflow code provenance tracking.
+"""
+
+import os
+import subprocess
+from os import path, name, environ
+from typing import Dict, Union
+
+# Cache for git information to avoid repeated subprocess calls
+_git_info_cache = None
+
+__all__ = ("get_repository_info",)
+
+GIT_COMMAND = "git"
+
+
+def _get_repo_url(path: Union[str, os.PathLike]):
+    """Get the repository URL from git config"""
+    try:
+        result = subprocess.run(
+            [GIT_COMMAND, "config", "--get", "remote.origin.url"],
+            cwd=path,
+            capture_output=True,
+            text=True,
+            check=False,
+        )
+        if result.returncode == 0:
+            url = result.stdout.strip()
+            # Convert SSH URLs to HTTPS for clickable links
+            if url.startswith("git@"):
+                parts = url.split(":")
+                if len(parts) == 2:
+                    domain = parts[0].replace("git@", "")
+                    repo_path = parts[1]
+                    url = f"https://{domain}/{repo_path}"
+            return url
+        return None
+    except (OSError, subprocess.SubprocessError):
+        return None
+
+
+def _get_branch_name(path: Union[str, os.PathLike]):
+    """Get the current git branch name"""
+    try:
+        result = subprocess.run(
+            [GIT_COMMAND, "rev-parse", "--abbrev-ref", "HEAD"],
+            cwd=path,
+            capture_output=True,
+            text=True,
+            check=False,
+        )
+        return result.stdout.strip() if result.returncode == 0 else None
+    except (OSError, subprocess.SubprocessError):
+        return None
+
+
+def _get_commit_sha(path: Union[str, os.PathLike]):
+    """Get the current git commit SHA"""
+    try:
+        result = subprocess.run(
+            [GIT_COMMAND, "rev-parse", "HEAD"],
+            cwd=path,
+            capture_output=True,
+            text=True,
+            check=False,
+        )
+        return result.stdout.strip() if result.returncode == 0 else None
```

Comment:
How do we distinguish between an error and non zero return code?

Reply:
we dont need to block the execution if somehow git metadata collection fails - it is strictly best effort by design.